api/src/main/java/com/netflix/iceberg/io/FileIO.java

```diff
@@ -47,4 +46,30 @@
    * Delete the file at the given path.
    */
   void deleteFile(String path);
+
+  /**
+   * Get an {@link InputFile} to get the bytes for this table's metadata file with the given name.
+   */
+  InputFile readMetadataFile(String fileName);
```

Comment:
Metastore tables always use full paths to metadata files, so this is only used by file system tables in `HadoopTables`. I think the need for this method is caused by not separating the concern of building a metadata path from getting an `InputFile` or `OutputFile` instance. It seems odd to me to add a method to this interface just for Hadoop tables. Another minor issue is that it isn't obvious why this takes a filename and not a full path: the name `readMetadataFile` doesn't indicate that to me. If a caller has a full path to a metadata file, why indicates that `readMetadataFile` is not the right choice?

On the other hand, if this were replaced by `metadataFileLocation`, it would still only be used by Hadoop tables because it is built into `newMetadataFile`. I like the way `newMetadataFile` and `newPartitionedDataFile` work, so I'd rather not change those.

What about removing `readMetadataFile` and instead using `newMetadataFile(filename).toInputFile()`? I think that it is clear that `newMetadataFile` adds the metadata path because it is distinct from `newOutputFile` and how to construct the full file path is a concern. I think this is a reasonable way to "guess" the location of metadata files for Hadoop tables.